### ISSUE TITLE
importadded plugin don't seem to work when importing from a zip file

### ISSUE THREAD

**Issue by zuzzurro** [beetbox/beets#4392]
Importing a disc from a zip file while having the importadded set results in file times not corresponding to the content of the zip archive itself, the files times are current. Repeating the same import on a folder extracted from the zip file using the unzip tool works fine.

I'm using 1.6.0 on Fedora 36 (self compiled) as the distro still has 1.4.9.

I can provide more information, but the issue is quite easy to reproduce as I explained above.

**Comment by sampsyo** [comment]
Interesting! I think the first step to understanding this would be to know whether the library we use to extract these files, namely [the `zipfile.extractall` function](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.extractall), can preserve mtimes. If it does, then we would need to find where we're discarding those. If it doesn't, then we may be out of luck. Any chance you'd be able to investigate?

**Comment by zuzzurro** [comment]
It looks like if people want to extract the correct metadata it needs to be done manually afterwards. Maybe similar to this?

[https://www.burgundywall.com/post/preserving-file-perms-with-python-zipfile-module](url)

**Comment by arogl** [comment]
I found the following, so it seems a limitation of pythons implementation of zip extraction:

https://stackoverflow.com/questions/9813243/extract-files-from-zip-file-and-retain-mod-date

**Comment by zuzzurro** [comment]
Would it be possible to adopt one of the proposed solutions? Since it's for internal use I don't see much possibly going wrong by changing the time by hand after the extraction.

**Comment by sampsyo** [comment]
Nice find, @arogl! It seems technically possible, but somewhat annoying to implement because we can no longer just use that `extractall` function… but I'll mark this as a feature request in case anyone is interested in giving it a shot.

**Comment by zuzzurro** [comment]
I haven't tested if the functionality works for RAR files, but since I think the corresponding functionality is implemented by directly calling an external utility (unrar) maybe that's the case.

**Comment by arogl** [comment]
@sampsyo 

Could this work?

In [importer.py#L1080](https://github.com/beetbox/beets/blob/master/beets/importer.py#L1080) add

``` py

# From here:
# https://stackoverflow.com/questions/9813243/extract-files-from-zip-file-and-retain-mod-date
# fixing #4392

def RestoreTimestampsOfArchiveContents(archivename, extract_dir):
    for f in archivename.infolist():
        # path to this extracted f-item
        fullpath = os.path.join(extract_dir, f.filename)
        # still need to adjust the dt o/w item will have the current dt
        date_time = time.mktime(f.date_time + (0, 0, -1))
        # update dt
        os.utime(fullpath, (date_time, date_time))
```

Then at [importer.py#L1093](https://github.com/beetbox/beets/blob/master/beets/importer.py#L1093):

```py
if (config['preserve_mtimes'].get(bool)):
    RestoreTimestampsOfArchiveContents(archive, extract_to)

```

I have not thought too much about PY2 example v. PY3, nor the if config

**Comment by sampsyo** [comment]
Yes, something like this could work! With the caveat that the `preserve_mtimes` option is located within the `importadded` configuration—not at the top level of `config`.

**Comment by zuzzurro** [comment]
By doing it this way are we setting the times also in the cases where they are already set by the unarchiver? Just curious.

**Comment by arogl** [comment]
> By doing it this way are we setting the times also in the cases where they are already set by the unarchiver? Just curious.

At the moment every extraction, regardless of type.

Further testing to be done

**Comment by arogl** [comment]
Initial change pushed #4396
